src/src/com/microsoft/aad/adal/IdToken.java

```diff
@@ -23,28 +23,163 @@
-    String mSubject;
+import java.io.UnsupportedEncodingException;
+import java.util.HashMap;
+import java.util.Iterator;
 
-    String mTenantId;
+import org.json.JSONException;
+import org.json.JSONObject;
 
-    String mUpn;
+import android.util.Base64;
 
-    String mGivenName;
+public class IdToken {
+    
+    final private static String TAG = "IdToken";
+    
+    private String mSubject;
+
+    private String mTenantId;
+
+    private String mUpn;
+
+    private String mGivenName;
 
-    String mFamilyName;
+    private String mFamilyName;
 
-    String mEmail;
+    private String mEmail;
 
-    String mIdentityProvider;
+    private String mIdentityProvider;
     
-    String mObjectId;
+    private String mObjectId;
     
-    long mPasswordExpiration;
+    private long mPasswordExpiration;
     
-    String mPasswordChangeUrl;
-        
-    public IdToken(){
+    private String mPasswordChangeUrl;
+    
+    public IdToken(String idtoken) throws AuthenticationException {
+        // Message segments: Header.Body.Signature
+        HashMap<String, String> responseItems = this.extractBody(idtoken);
+            
+        if (responseItems != null && !responseItems.isEmpty()) {
+            this.mSubject = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_SUBJECT);
+            this.mTenantId = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_TENANTID);
+            this.mUpn = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_UPN);
+            this.mEmail = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_EMAIL);
+            this.mGivenName = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_GIVEN_NAME);
+            this.mFamilyName = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_FAMILY_NAME);
+            this.mIdentityProvider = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_IDENTITY_PROVIDER);
+            this.mObjectId = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_OBJECT_ID);
+            String expiration = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_PASSWORD_EXPIRATION);
+            
+            if (!StringExtensions.IsNullOrBlank(expiration)) {
+                this.mPasswordExpiration = Long.parseLong(expiration);
+            }
+            
+            this.mPasswordChangeUrl = responseItems
+                    .get(AuthenticationConstants.OAuth2.ID_TOKEN_PASSWORD_CHANGE_URL);
+        }
+    }
+    
+    private HashMap<String, String> extractBody(String idtoken) throws AuthenticationException {
+        int firstDot = idtoken.indexOf(".");
+        int secondDot = idtoken.indexOf(".", firstDot + 1);
+        int invalidDot = idtoken.indexOf(".", secondDot + 1);
+
+        if (invalidDot == -1 && firstDot > 0 && secondDot > 0) {
```

Comment:
I'm not sure that I understand the question.  the idToken is a standard JWT token.  A jwt token has two periods.  The format is <header>.<body>.<signature> this code is being extra cautious to check the schema and not bother parsing if there is an extra period.  Do you see a problem with that?